Content.Server/Body/Systems/InternalsSystem.cs

```diff
@@ -1,24 +1,175 @@
+            args.Gas = _gasTank.RemoveAirVolume(gasTank, Atmospherics.BreathVolume);
+            // TODO: Should listen to gas tank updates instead I guess?
+            _alerts.ShowAlert(component.Owner, AlertType.Internals, GetSeverity(component));
+        }
+    }
+    public void DisconnectBreathTool(InternalsComponent component)
+    {
+        var old = component.BreathToolEntity;
+        component.BreathToolEntity = null;
+
+        if (TryComp(old, out BreathToolComponent? breathTool) )
+        {
+            _atmos.DisconnectInternals(breathTool);
+            DisconnectTank(component);
+        }
+
+        _alerts.ShowAlert(component.Owner, AlertType.Internals, GetSeverity(component));
+    }
+
+    public void ConnectBreathTool(InternalsComponent component, EntityUid toolEntity)
+    {
+        if (TryComp(component.BreathToolEntity, out BreathToolComponent? tool))
+        {
+            _atmos.DisconnectInternals(tool);
+        }
+
+        component.BreathToolEntity = toolEntity;
+        _alerts.ShowAlert(component.Owner, AlertType.Internals, GetSeverity(component));
+    }
+
+    public void DisconnectTank(InternalsComponent? component)
+    {
+        if (component == null) return;
+
+        if (TryComp(component.GasTankEntity, out GasTankComponent? tank))
+        {
+            _gasTank.DisconnectFromInternals(tank, component.Owner);
+        }
+
+        component.GasTankEntity = null;
+        _alerts.ShowAlert(component.Owner, AlertType.Internals, GetSeverity(component));
+    }
+
+    public bool TryConnectTank(InternalsComponent component, EntityUid tankEntity)
+    {
+        if (component.BreathToolEntity == null)
+            return false;
+
+        if (TryComp(component.GasTankEntity, out GasTankComponent? tank))
+        {
+            _gasTank.DisconnectFromInternals(tank, component.Owner);
+        }
+
+        component.GasTankEntity = tankEntity;
+        _alerts.ShowAlert(component.Owner, AlertType.Internals, GetSeverity(component));
+        return true;
+    }
+
+    public bool AreInternalsWorking(InternalsComponent component)
+    {
+        return TryComp(component.BreathToolEntity, out BreathToolComponent? breathTool) &&
+               breathTool.IsFunctional &&
+               TryComp(component.GasTankEntity, out GasTankComponent? gasTank) &&
+               gasTank.Air != null;
+    }
+
+    private short GetSeverity(InternalsComponent component)
+    {
+        if (component.BreathToolEntity == null || !AreInternalsWorking(component)) return 2;
+
+        if (TryComp<GasTankComponent>(component.GasTankEntity, out var gasTank) && gasTank.Air.Volume < Atmospherics.BreathVolume)
+            return 0;
+
+        return 1;
+    }
+
+    public GasTankComponent? FindBestGasTank(InternalsComponent component)
+    {
+        // Prioritise
+        // 1. exo-slot tanks
+        // 2. in-hand tanks
+        // 3. pocket tanks
+        InventoryComponent? inventory = null;
```

Comment:
Eventually inventory relay events should somehow support slot priorities & checking hands, but I guess its fine to do it like this for now.